community/kernel/src/test/java/org/neo4j/kernel/impl/locking/community/RWLockTest.java

```diff
@@ -0,0 +1,456 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.locking.community;
+
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import javax.transaction.Transaction;
+
+import org.neo4j.kernel.DeadlockDetectedException;
+import org.neo4j.kernel.impl.locking.ResourceTypes;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+public class RWLockTest
+{
+
+    private static ExecutorService executor;
+
+    @BeforeClass
+    public static void initExecutor()
+    {
+        executor = Executors.newCachedThreadPool();
+    }
+
+    @AfterClass
+    public static void stopExecutor() throws InterruptedException
+    {
+        executor.shutdown();
+        executor.awaitTermination( 2, TimeUnit.SECONDS );
+    }
+
+    @Test
+    public void assertWriteLockDoesNotLeakMemory() throws InterruptedException
+    {
+        final RagManager ragManager = new RagManager();
+        final Object resource = new Object();
+        final RWLock lock = new RWLock( resource, ragManager );
+        final Transaction tx1 = mock( Transaction.class );
+
+        lock.mark();
+        lock.acquireWriteLock( tx1 );
+        lock.mark();
+
+        assertEquals( 1, lock.getTxLockElementCount() );
+        lock.releaseWriteLock( tx1 );
+        assertEquals( 0, lock.getTxLockElementCount() );
+    }
+
+    @Test
+    public void assertReadLockDoesNotLeakMemory() throws InterruptedException
+    {
+        final RagManager ragManager = new RagManager();
+        final Object resource = new Object();
+        final RWLock lock = new RWLock( resource, ragManager );
+        final Transaction tx1 = mock( Transaction.class );
+
+        lock.mark();
+        lock.acquireReadLock( tx1 );
+        lock.mark();
+
+        assertEquals( 1, lock.getTxLockElementCount() );
+        lock.releaseReadLock( tx1 );
+        assertEquals( 0, lock.getTxLockElementCount() );
+    }
+
+    /*
+     * In case if writer thread can't grab write lock now, it should be added to
+     * into a waiting list, wait till resource will be free and grab it.
+     */
+    @Test( timeout = 500 )
```

Comment:
I think you should increase the timeout to at least 1 second, because waking up a thread on an EC2 virtual machine can sometimes take a surprisingly long time.